File: Daily/645. Set Mismatch/Solution3.py
from typing import List
from collections import Counter, defaultdict
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        '''
            Optimized Solution using Maps
            Time Complexity = O(n)
            Space Complexity = O(n)
        '''
        total_Elements, duplicate, missing = len(nums), -1, -1
        # frequency = dict(Counter(nums))
        frequency = defaultdict(int)
        for element in nums: frequency[element] += 1
        for element in range(1, total_Elements + 1):
            if frequency.get(element, 0) == 0: missing = element
            elif frequency[element] > 1: duplicate = element
        return [duplicate, missing]

if __name__ == '__main__':
    nums = [int(x) for x in input().split()]; sol = Solution()
    ans = sol.findErrorNums(nums)
    print('Duplicate Element =', ans[0], 'and Missing Element =', ans[1])

File: Daily/645. Set Mismatch/Solution3.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;
class Solution {
    public:
        vector<int> findErrorNums(vector<int>& nums) {
            /*
                Optimized Solution using Maps
                Time Complexity = O(n)
                Space Complexity = O(n)
            */
            unordered_map<int, int> frequency;
            int duplicate = -1, missing = -1; const int total_Elements = nums.size();
            for(const int& element: nums) frequency[element]++;
            for(int element = 1; element <= total_Elements; ++element)
                if(frequency.find(element) == frequency.end()) missing = element;
                else if(frequency[element] > 1) duplicate = element;
            return {duplicate, missing};
        }
};
int main() {
    int n; cin >> n; vector<int> nums(n);
    for(int index = 0; index < n; index++) cin >> nums[index];
    Solution sol; vector<int> ans = sol.findErrorNums(nums);
    cout << "Duplicate Number = " << ans[0] 
        << " and Missing Number = " << ans[1] << endl;
}